File: src/cpp/http/response.cpp
#include "response.h"
#include "error.h"
#include "sdk-sys.h"

namespace fastly::http {

Response::Response()
    : res(std::move(fastly::sys::http::m_static_http_response_new())) {}

bool Response::is_from_backend() { return this->res->is_from_backend(); }

Response Response::clone_without_body() {
  Response res{this->res->clone_without_body()};
  return res;
}

Response Response::clone_with_body() {
  Response res{this->res->clone_with_body()};
  return res;
}

Response Response::from_body(Body body) {
  Response res(std::move(fastly::sys::http::m_static_http_response_from_body(
      std::move(body.bod))));
  return res;
}

Response Response::from_status(StatusCode status) {
  Response res(std::move(
      fastly::sys::http::m_static_http_response_from_status(status.as_code())));
  return res;
}

Response Response::with_body(Body body) {
  this->set_body(std::move(body));
  return std::move(*this);
}

void Response::set_body(Body body) {
  // TODO(@zkat): this is broken rn because I messed up the move constructor for
  // Body/StreamingBody.
  body.flush();
  this->res->set_body(std::move(body.bod));
}

Body Response::take_body() {
  Body body{this->res->take_body()};
  return body;
}

void Response::append_body(Body other) {
  other.flush();
  this->res->append_body(std::move(other.bod));
}

std::vector<uint8_t> Response::into_body_bytes() {
  auto str{this->into_body_string()};
  return std::vector<uint8_t>(str.begin(), str.end());
}

std::string Response::into_body_string() {
  Body body{this->into_body()};
  return std::string(std::istreambuf_iterator<char>(body),
                     std::istreambuf_iterator<char>());
}

Body Response::into_body() {
  return Body(
      fastly::sys::http::m_http_response_into_body(std::move(this->res)));
}

fastly::expected<void> Response::set_body_text_plain(std::string_view body) {
  fastly::sys::error::FastlyError *err;
  this->res->set_body_text_plain(static_cast<std::string>(body), err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return fastly::expected<void>();
  }
}

fastly::expected<Response>
Response::with_body_text_html(std::string_view body) {
  return this->set_body_text_html(body).map(
      [this]() { return std::move(*this); });
}

fastly::expected<void> Response::set_body_text_html(std::string_view body) {
  fastly::sys::error::FastlyError *err;
  this->res->set_body_text_html(static_cast<std::string>(body), err);
  if (err != nullptr) {
    return fastly::expected<void>();
  } else {
    return fastly::unexpected(err);
  }
}

fastly::expected<Response>
Response::with_body_text_plain(std::string_view body) {
  return this->set_body_text_plain(body).map(
      [this]() { return std::move(*this); });
}

std::string Response::take_body_string() {
  Body bod{this->take_body()};
  return std::string(std::istreambuf_iterator<char>(bod),
                     std::istreambuf_iterator<char>());
}

Response Response::with_body_octet_stream(std::vector<uint8_t> body) {
  this->set_body_octet_stream(body);
  return std::move(*this);
}

void Response::set_body_octet_stream(std::vector<uint8_t> body) {
  this->res->set_body_octet_stream(body);
}

std::vector<uint8_t> Response::take_body_bytes() {
  auto body_str{this->take_body_string()};
  return std::vector<uint8_t>(body_str.begin(), body_str.end());
}

// ChunksIter Response::read_body_chunks(size_t chunk_size);

std::optional<std::string> Response::get_content_type() {
  std::string out;
  if (this->res->get_content_type(out)) {
    return out;
  } else {
    return std::nullopt;
  }
}

Response Response::with_content_type(std::string_view mime) {
  this->set_content_type(mime);
  return std::move(*this);
}

void Response::set_content_type(std::string_view mime) {
  this->res->set_content_type(static_cast<std::string>(mime));
}

std::optional<size_t> Response::get_content_length() {
  size_t len;
  if (this->res->get_content_length(len)) {
    return len;
  } else {
    return std::nullopt;
  }
}

fastly::expected<bool> Response::contains_header(std::string_view name) {
  fastly::sys::error::FastlyError *err;
  bool has_header{
      this->res->contains_header(static_cast<std::string>(name), err)};
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return fastly::expected<bool>(has_header);
  }
}

fastly::expected<Response> Response::with_header(std::string_view name,
                                                 std::string_view value) {
  return this->append_header(name, value).map([this]() {
    return std::move(*this);
  });
}

fastly::expected<Response> Response::with_set_header(std::string_view name,
                                                     std::string_view value) {
  return this->set_header(name, value).map([this]() {
    return std::move(*this);
  });
}

// TODO(@zkat): do a proper HeaderValue situation here?
fastly::expected<std::optional<std::string>>
Response::get_header(std::string_view name) {
  fastly::sys::error::FastlyError *err;
  std::string out;
  bool has_header{
      this->res->get_header(static_cast<std::string>(name), out, err)};
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else if (has_header) {
    return std::optional<std::string>(std::move(out));
  } else {
    return std::nullopt;
  }
}

fastly::expected<HeaderValuesIter>
Response::get_header_all(std::string_view name) {
  fastly::sys::http::HeaderValuesIter *out;
  fastly::sys::error::FastlyError *err;
  this->res->get_header_all(static_cast<std::string>(name), out, err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return FSLY_BOX(http, HeaderValuesIter, out);
  }
}

// TODO(@zkat): sigh. IDK
// ??? get_headers();
// HeaderNamesIter get_header_names();

fastly::expected<void> Response::set_header(std::string_view name,
                                            std::string_view value) {
  fastly::sys::error::FastlyError *err;
  this->res->set_header(static_cast<std::string>(name),
                        static_cast<std::string>(value), err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return fastly::expected<void>();
  }
}

fastly::expected<void> Response::append_header(std::string_view name,
                                               std::string_view value) {
  fastly::sys::error::FastlyError *err;
  this->res->append_header(static_cast<std::string>(name),
                           static_cast<std::string>(value), err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return fastly::expected<void>();
  }
}

fastly::expected<std::optional<std::string>>
Response::remove_header(std::string_view name) {
  fastly::sys::error::FastlyError *err;
  std::string out;
  bool has_header{
      this->res->remove_header(static_cast<std::string>(name), out, err)};
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else if (has_header) {
    return std::optional<std::string>(std::move(out));
  } else {
    return std::nullopt;
  }
}

void Response::set_status(StatusCode status) {
  this->res->set_status(status.as_code());
}

Response Response::with_status(StatusCode status) {
  this->set_status(status);
  return std::move(*this);
}

std::optional<std::string> Response::get_backend_name() {
  std::string name;
  bool existed{this->res->get_backend_name(name)};
  if (!existed) {
    return std::nullopt;
  } else {
    return {name};
  }
}

std::optional<fastly::backend::Backend> Response::get_backend() {
  auto ptr{this->res->get_backend()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return FSLY_BOX(backend, Backend, ptr);
  }
}

std::optional<std::string> Response::get_backend_addr() {
  std::string addr;
  bool existed{this->res->get_backend_name(addr)};
  if (!existed) {
    return std::nullopt;
  } else {
    return {addr};
  }
}

std::optional<Request> Response::take_backend_request() {
  auto ptr{this->res->take_backend_request()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return FSLY_BOX(http, Request, ptr);
  }
}

void Response::send_to_client() {
  // TODO(@zkat): flush body before sending.
  fastly::sys::http::m_http_response_send_to_client(std::move(this->res));
}

StreamingBody Response::stream_to_client() {
  return {fastly::sys::http::m_http_response_stream_to_client(
      std::move(this->res))};
}

std::optional<std::chrono::milliseconds> Response::get_ttl() {
  uint32_t out;
  if (this->res->get_ttl(out)) {
    return std::chrono::milliseconds(out);
  } else {
    return std::nullopt;
  }
}

std::optional<std::chrono::milliseconds> Response::get_age() {
  uint32_t out;
  if (this->res->get_age(out)) {
    return std::chrono::milliseconds(out);
  } else {
    return std::nullopt;
  }
}

std::optional<std::chrono::milliseconds>
Response::get_stale_while_revalidate() {
  uint32_t out;
  if (this->res->get_stale_while_revalidate(out)) {
    return std::chrono::milliseconds(out);
  } else {
    return std::nullopt;
  }
}

} // namespace fastly::http

File: src/http/response.rs
use std::pin::Pin;

use cxx::{CxxString, CxxVector, UniquePtr};
use http::{HeaderName, HeaderValue};

use crate::{
    backend::Backend,
    error::ErrPtr,
    http::{
        body::{Body, StreamingBody},
        header::HeaderValuesIter,
        request::Request,
    },
    try_fe,
};

pub struct Response(pub(crate) fastly::Response);

pub fn m_static_http_response_new() -> Box<Response> {
    Box::new(Response(fastly::Response::new()))
}

pub fn m_http_response_send_to_client(response: Box<Response>) {
    response.0.send_to_client();
}

pub fn m_http_response_stream_to_client(response: Box<Response>) -> Box<StreamingBody> {
    Box::new(StreamingBody(response.0.stream_to_client()))
}

pub fn m_static_http_response_from_body(body: Box<Body>) -> Box<Response> {
    Box::new(Response(fastly::Response::from_body(body.0)))
}

pub fn m_static_http_response_from_status(status: u16) -> Box<Response> {
    Box::new(Response(fastly::Response::from_status(status)))
}

pub fn m_static_http_response_see_other(destination: &CxxString) -> Box<Response> {
    Box::new(Response(fastly::Response::see_other(
        destination.as_bytes(),
    )))
}

pub fn m_static_http_response_redirect(destination: &CxxString) -> Box<Response> {
    Box::new(Response(fastly::Response::redirect(destination.as_bytes())))
}

pub fn m_static_http_response_temporary_redirect(destination: &CxxString) -> Box<Response> {
    Box::new(Response(fastly::Response::temporary_redirect(
        destination.as_bytes(),
    )))
}

pub fn m_http_response_into_body(response: Box<Response>) -> Box<Body> {
    Box::new(Body(response.0.into_body()))
}

impl Response {
    pub fn is_from_backend(&self) -> bool {
        self.0.is_from_backend()
    }

    pub fn clone_without_body(&self) -> Box<Response> {
        Box::new(Response(self.0.clone_without_body()))
    }

    pub fn clone_with_body(&mut self) -> Box<Response> {
        Box::new(Response(self.0.clone_with_body()))
    }

    pub fn has_body(&self) -> bool {
        self.0.has_body()
    }

    pub fn set_body(&mut self, body: Box<Body>) {
        self.0.set_body(body.0);
    }

    pub fn take_body(&mut self) -> Box<Body> {
        Box::new(Body(self.0.take_body()))
    }

    pub fn append_body(&mut self, other: Box<Body>) {
        self.0.append_body(other.0);
    }

    pub fn set_body_text_plain(&mut self, body: &CxxString, mut err: ErrPtr) {
        self.0.set_body_text_plain(try_fe!(err, body.to_str()));
    }

    pub fn set_body_text_html(&mut self, body: &CxxString, mut err: ErrPtr) {
        self.0.set_body_text_html(try_fe!(err, body.to_str()));
    }

    pub fn set_body_octet_stream(&mut self, body: &CxxVector<u8>) {
        self.0.set_body_octet_stream(body.as_slice())
    }

    pub fn get_content_type(&self, out: Pin<&mut CxxString>) -> bool {
        self.0
            .get_content_type()
            .map(|mime| out.push_str(mime.as_ref()))
            .is_some()
    }

    pub fn set_content_type(&mut self, mime: &CxxString) {
        self.0.set_content_type(
            mime.to_str()
                .expect("Invalid UTF-8")
                .parse()
                .expect("Invalid MIME type"),
        )
    }

    pub fn get_content_length(&self, mut out: Pin<&mut usize>) -> bool {
        self.0
            .get_content_length()
            .map(|len| out.set(len))
            .is_some()
    }

    pub fn contains_header(&self, name: &CxxString, mut err: ErrPtr) -> bool {
        self.0
            .contains_header(try_fe!(err, HeaderName::try_from(name.as_bytes())))
    }

    pub fn get_header(&self, name: &CxxString, out: Pin<&mut CxxString>, mut err: ErrPtr) -> bool {
        self.0
            .get_header(try_fe!(err, HeaderName::try_from(name.as_bytes())))
            .map(|header| out.push_bytes(header.as_bytes()))
            .is_some()
    }

    pub fn get_header_all(
        &self,
        name: &CxxString,
        mut out: Pin<&mut *mut HeaderValuesIter>,
        mut err: ErrPtr,
    ) {
        // Yeah. Sorry. Lifetimes :/
        let iter = self
            .0
            .get_header_all(try_fe!(err, HeaderName::try_from(name.as_bytes())))
            .map(|v| {
                let mut vector = CxxVector::new();
                for byte in v.as_bytes() {
                    vector.pin_mut().push(*byte);
                }
                vector
            })
            .collect::<Vec<UniquePtr<CxxVector<u8>>>>();
        out.set(Box::into_raw(Box::new(HeaderValuesIter(Box::new(
            iter.into_iter(),
        )))))
    }

    pub fn set_header(&mut self, name: &CxxString, value: &CxxString, mut err: ErrPtr) {
        self.0.set_header(
            try_fe!(err, HeaderName::try_from(name.as_bytes())),
            try_fe!(err, HeaderValue::try_from(value.as_bytes())),
        );
    }

    pub fn append_header(&mut self, name: &CxxString, value: &CxxString, mut err: ErrPtr) {
        self.0.append_header(
            try_fe!(err, HeaderName::try_from(name.as_bytes())),
            try_fe!(err, HeaderValue::try_from(value.as_bytes())),
        );
    }

    pub fn remove_header(
        &mut self,
        name: &CxxString,
        out: Pin<&mut CxxString>,
        mut err: ErrPtr,
    ) -> bool {
        self.0
            .remove_header(try_fe!(err, HeaderName::try_from(name.as_bytes())))
            .map(|header| out.push_bytes(header.as_bytes()))
            .is_some()
    }

    pub fn set_status(&mut self, status: u16) {
        self.0.set_status(status);
    }

    pub fn get_backend_name(&self, out: Pin<&mut CxxString>) -> bool {
        self.0.get_backend_name().map(|b| out.push_str(b)).is_some()
    }

    pub fn get_backend(&self) -> *mut Backend {
        if let Some(b) = self.0.get_backend() {
            Box::into_raw(Box::new(Backend(b.clone())))
        } else {
            std::ptr::null_mut()
        }
    }

    pub fn get_backend_addr(&self, out: Pin<&mut CxxString>) -> bool {
        self.0
            .get_backend_addr()
            .map(|ip| out.push_str(ip.to_string().as_ref()))
            .is_some()
    }

    pub fn take_backend_request(&mut self) -> *mut Request {
        if let Some(req) = self.0.take_backend_request() {
            Box::into_raw(Box::new(Request(req)))
        } else {
            std::ptr::null_mut()
        }
    }

    pub fn get_ttl(&self, mut out: Pin<&mut u32>) -> bool {
        self.0
            .get_ttl()
            .map(|ttl| ttl.as_millis())
            .map(ensure_u32)
            .map(|ttl| out.set(ttl))
            .is_some()
    }

    pub fn get_age(&self, mut out: Pin<&mut u32>) -> bool {
        self.0
            .get_age()
            .map(|age| age.as_millis())
            .map(ensure_u32)
            .map(|age| out.set(age))
            .is_some()
    }

    pub fn get_stale_while_revalidate(&self, mut out: Pin<&mut u32>) -> bool {
        self.0
            .get_age()
            .map(|swr| swr.as_millis())
            .map(ensure_u32)
            .map(|swr| out.set(swr))
            .is_some()
    }
}

fn ensure_u32(num: u128) -> u32 {
    if num > std::u32::MAX as u128 {
        std::u32::MAX
    } else {
        num as u32
    }
}
